# *IS 362 - Project 1*

## Importing Data (CSV)

In [1]:
import pandas as pd

In [2]:
# This is to help identify a filepath, but it is not needed for this scenario
from pathlib import Path
print(Path.cwd())

/home/cf23896372/Documents


As opposed to creating a dataframe from scratch, extracting data from a prexisting file such as a spreadsheet (e.g., Excel) or database (e.g., SQL, JSON, CSV, etc.) would result some headers and row spaces to have values such as "Unnamed" or "NaN" for cells/spaces that are intentionally left blank in the original file. To indicate labels of row data, they would have be indicated as index row using "index_col=[**the column numbers needed**]."

In [3]:
# Had the filepath method been used, this string of code below would be needed
# location = r'/home/cf23896372/Documents'

# But we don't need it, so this block of code is sufficient
airtime = pd.read_csv('Airline_Data.csv', index_col=[0, 1])
airtime

Los Angeles  Phoenix  San Diego  San Francisco  Seattle
ALASKA  on time          497      221        212            503     1841
        delayed           62       12         20            102      305
AM WEST on time          694     4840        383            320      201
        delayed          117      415         65            129       61

## Testing Dataframe

These cells below are to help check the data format and structure, ensuring the original CSV structure was transferred when imported into Jupyter. Apparently, since two of the rows had the same airline name in the cells, the two cells merged in the pandas dataframe output, indicating that they each have "on time" and "delayed" row of data for each airline. 

In [4]:
print(airtime)

                 Los Angeles  Phoenix  San Diego  San Francisco  Seattle
ALASKA  on time          497      221        212            503     1841
        delayed           62       12         20            102      305
AM WEST on time          694     4840        383            320      201
        delayed          117      415         65            129       61


In [5]:
# How many rows of data are there?
len(airtime)

4

In [6]:
# Retrieves index data
airtime.index

MultiIndex([( 'ALASKA', 'on time'),
            ( 'ALASKA', 'delayed'),
            ('AM WEST', 'on time'),
            ('AM WEST', 'delayed')],
           )

## Data Analyses

In [7]:
flights_by_city = airtime.sum()

city_flights_sum = pd.DataFrame(flights_by_city)
city_flights_sum.columns = ['Flights Arrived']
city_flights_sum

Flights Arrived
Los Angeles               1370
Phoenix                   5488
San Diego                  680
San Francisco             1054
Seattle                   2408

This was a practice to showcase the number of flights that have arrived in each city in total, regardless whether or not their arrival was delayed. As shown on the dataframe, Phoenix, AZ had the most amount of flights arrive there, while San Diego, CA had the least.

In [8]:
# Retrives the sum of all rows
airtime_row_sum = airtime.sum(axis=1)

data = airtime_row_sum

flights_sum = pd.DataFrame(data)
# Retrieves an already retrieved multi-index
flights_sum.index = airtime.index
# Gives name to blank column header
flights_sum.columns = ['Total Number of Flights By Arrival']

flights_sum

Total Number of Flights By Arrival
ALASKA  on time                                3274
        delayed                                 501
AM WEST on time                                6438
        delayed                                 787

While this is good to display the total number of flights for each airline individually, as well as categorizing them by whether the flights' arrival was on time or delayed, this will not be suitable for a comparison between the two airlines. The sample size between the two airlines are non-proportional to one another, especially since America West Airlines (**AM WEST**) has a larger sample size compared to Alaska Airlines (**ALASKA**). Thus, I need to calculate the ratio of number of both on-time and delayed arrivals to the total number of arrivals of each airline. But first, the sum of the total number of arrivals for each airline must be calculated, giving the observational sample size of both airlines. Afterwards, I can calculate the on-time and delayed arrival rates for each airline.

In [9]:
# The code below will sum the values of multiple rows specified by index
ALASKA_total_arrivals = flights_sum.loc['ALASKA'].sum()
AMWest_total_arrivals = flights_sum.loc['AM WEST'].sum()

arrival_data = [ALASKA_total_arrivals, AMWest_total_arrivals]

arrival_sum = pd.DataFrame(arrival_data)
# I do not know how to get a specific index pulled from a multi-index, so I had to resort to this
arrival_sum.index = ['ALASKA', 'AM WEST']
arrival_sum.columns = ['Total Number of Arrivals']

arrival_sum

Total Number of Arrivals
ALASKA                       3775
AM WEST                      7225

The sum of the flight data from each airline in the dataframe calculates the number of overall flights that have arrived on time or delayed, not factoring in the location of arrival. However, while it can be argued that both airlines have more on time arrivals than delayed arrivals, the sum of the flights in total for each airline are different, with America West Airlines (**AM WEST**) having the most amount of flights compared to Alaska Airlines (**ALASKA**).

In [10]:
ALASKA_OnTime = flights_sum.loc['ALASKA', 'on time']['Total Number of Flights By Arrival']
ALASKA_Delayed = flights_sum.loc['ALASKA', 'delayed']['Total Number of Flights By Arrival']
ALASKA_Sum = arrival_sum.loc['ALASKA']['Total Number of Arrivals']

AMWest_OnTime = flights_sum.loc['AM WEST', 'on time']['Total Number of Flights By Arrival']
AMWest_Delayed = flights_sum.loc['AM WEST', 'delayed']['Total Number of Flights By Arrival']
AMWest_Sum = arrival_sum.loc['AM WEST']['Total Number of Arrivals']

# The value of the ratio is multiplied by 100 for it to be a percentage
ratio1 = (ALASKA_OnTime/ALASKA_Sum)*100
ratio2 = (ALASKA_Delayed/ALASKA_Sum)*100
ratio3 = (AMWest_OnTime/AMWest_Sum)*100
ratio4 = (AMWest_Delayed/AMWest_Sum)*100

ratio_data = [ratio1, ratio2, ratio3, ratio4]

percentage_flight_data = pd.DataFrame(ratio_data)
percentage_flight_data.index = airtime.index
percentage_flight_data.columns = ['Overall Flight Arrival Rate (%)']

percentage_flight_data

Overall Flight Arrival Rate (%)
ALASKA  on time                        86.728477
        delayed                        13.271523
AM WEST on time                        89.107266
        delayed                        10.892734

As shown in the dataframe, Alaska Airlines (**ALASKA**) is shown to have the most amount of delayed flights in comparison to America West Airlines (**AM WEST**),its rate having approximately 13.27% of its flight arrivals delayed compared to American West Airlines' flights' rate of having only approximately 10.9% of flight arrival delays. 

Conversely, America West Airlines had more on time arrivals in comparison to Alaska Airlines, given that a rate of about 89.1% of America West Flights had on time arrivals compared to Alaska Airlines, in which about 86.73% of its flights arrived on time.

The content of the dataframe displays that America West Airlines has a larger on-time arrival rate and lower arrival delay rate compared to Alaska Airlines.

## Conclusion

As displayed by the data provided and the ratio of flight arrival performance to the total number of flight arrivals by each airline, it can be determined that America West Airlines (**AM WEST**) has a better flight arrival performance/schedule compared to Alaska Airlines (**ALASKA**), since the former has a higher on-time arrival percentage rate compared to the latter. In contrast, Alaska Airlines has a higher percentage rate of delays than America West Airlines.

This indicates that Alaska Airlines has a poor schedule performance/management as it has a higher arrival delay rate compared to American West Airlines. Thus, it can be concluded that America West Airlines has better schedule management given that it has a high on-time arrival rate, it can be predicted that the flights having a higher likelihood of arriving on time to its destination.

However, it should be noted that the accuracy of this conclusion is ambiguous, given the vastly different sample size provided in the observation data. Since the sample size of America West Airlines is significantly larger than Alaska Airlines', the comparison analysis between the two observations of airline arrival data may not be considered proper.

A better observational analysis can be made if the sample sizes were approximately the same, having roughly the same amount of flights for each airline (or both numbers are in close proximity to one another), creating a small margin of error and more accurate calculations of the flight arrival rates.